File: codeql-custom-queries-java/android/prefer-bluetooth-low-energy/prefer-bluetooth-low-energy.java
class BluetoothAdapter {
    public static BluetoothAdapter getDefaultAdapter() { return null; }
    public BluetoothDevice getRemoteDevice(String address) { return null; }
    public void startDiscovery() {}
    public void cancelDiscovery() {}
}

class BluetoothDevice {
    public BluetoothSocket createRfcommSocketToServiceRecord(java.util.UUID uuid) { return null; }
}

class BluetoothSocket {
    public void connect() {}
    public void close() {}
}

class BluetoothServerSocket {
    public BluetoothSocket accept() { return null; }
    public void close() {}
}

interface BluetoothProfile {
    int getConnectionState(BluetoothDevice device);
}

class NoncompliantBluetooth {

    // 🚫 Noncompliant - BluetoothAdapter field
    BluetoothAdapter adapter = null; // $ Alert

    // 🚫 Noncompliant - BluetoothDevice field
    BluetoothDevice device = null; // $ Alert

    // 🚫 Noncompliant - BluetoothSocket field
    BluetoothSocket socket = null; // $ Alert

    // 🚫 Noncompliant - BluetoothServerSocket field
    BluetoothServerSocket serverSocket = null; // $ Alert

    // 🚫 Noncompliant - BluetoothProfile field
    BluetoothProfile profile = null; // $ Alert

    void scanDevices() {
        // 🚫 Noncompliant - local BluetoothAdapter
        BluetoothAdapter localAdapter = BluetoothAdapter.getDefaultAdapter(); // $ Alert
        localAdapter.startDiscovery();
    }

    void connectToDevice(String address) {
        // 🚫 Noncompliant - local BluetoothDevice
        BluetoothAdapter a = BluetoothAdapter.getDefaultAdapter(); // $ Alert
        BluetoothDevice remoteDevice = a.getRemoteDevice(address); // $ Alert
    }

    void openSocket() throws Exception {
        // 🚫 Noncompliant - local BluetoothSocket
        BluetoothSocket s = null; // $ Alert
        s.connect();
    }

    void listenForConnections() throws Exception {
        // 🚫 Noncompliant - local BluetoothServerSocket
        BluetoothServerSocket ss = null; // $ Alert
        ss.accept();
    }
}

class CompliantBluetooth {
    // ✅ Compliant - no classic Bluetooth usage
    void doNothing() {}
}

File: codeql-custom-queries-java/android/prefer-bluetooth-low-energy.kt
// Stubs - classic Bluetooth
class BluetoothAdapter
class BluetoothDevice
class BluetoothSocket
class BluetoothServerSocket
class BluetoothProfile

// Stubs - Bluetooth Low Energy
class BluetoothLeScanner
class BluetoothLeAdvertiser
class ScanCallback
class ScanResult
class AdvertiseCallback

// 🚫 Noncompliant - using classic Bluetooth only
class NoncompliantBluetooth {

    val adapter: BluetoothAdapter? = null      // $ Alert
    val device: BluetoothDevice? = null        // $ Alert
    val socket: BluetoothSocket? = null        // $ Alert

    fun connect() {
        val serverSocket: BluetoothServerSocket? = null  // $ Alert
        val profile: BluetoothProfile? = null            // $ Alert
    }
}

// ✅ Compliant - using Bluetooth Low Energy
class CompliantBluetooth {

    val scanner: BluetoothLeScanner? = null       // OK
    val advertiser: BluetoothLeAdvertiser? = null // OK
    val callback: ScanCallback? = null            // OK
    val result: ScanResult? = null                // OK
}